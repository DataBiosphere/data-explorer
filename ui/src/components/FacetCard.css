.facetCard {
  margin: 2%;
  padding-bottom: 8px;
}

.facetValueName {
  float: left;
  text-align: left;
}

.facetValueCount {
  float: right;
  text-align: right;
}

.cardHeader {
  padding: 15px;
}

.subHeader {
  margin-top: 5px;
  color: gray;
}

.totalCount {
  float: left;
}

.numberSelected {
  float: right;
}

.grayText {
  color: silver !important;
}
