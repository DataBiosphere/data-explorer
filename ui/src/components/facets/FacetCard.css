.facetCard {
  margin: 2%;
  padding-bottom: 8px;
}

.cardHeader {
  padding-left: 15px;
  padding-right: 15px;
  padding-top: 15px;
}

.totalFacetValueCount {
  color: gray;
  float: right;
}

.facetDescription {
  color: gray;
}

.facetValueName {
  float: left;
  text-align: left;
}

.facetValueCount {
  float: right;
  text-align: right;
}

.grayText {
  color: silver !important;
}
