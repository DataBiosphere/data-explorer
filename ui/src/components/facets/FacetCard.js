import React, { Component } from "react";
import { withStyles } from "@material-ui/core/styles";
import Checkbox from "@material-ui/core/Checkbox";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import Typography from "@material-ui/core/Typography";

import * as Style from "libs/style";

const styles = {
  facetCard: {
    ...Style.elements.card,
    margin: "2%",
    paddingBottom: "8px",
    display: "grid",
    gridTemplateColumns: "auto 50px",
    // If there is a long word (eg facet name or facet value), break in the
    // middle of the word. Without this, the word stays on one line and its CSS
    // grid is wider than the facet card.
    wordBreak: "break-word"
  },
  // By default, each div takes up one grid cell.
  // Don't specify gridColumn, just use default of one cell.
  facetDescription: {
    color: "gray"
  },
  totalFacetValueCount: {
    color: "gray",
    textAlign: "right"
  },
  facetValueList: {
    gridColumn: "1 / 3",
    margin: "20px 0 0 0",
    maxHeight: "400px",
    overflow: "auto"
  },
  facetValue: {
    // This is a nested div, so need to specify a new grid.
    display: "grid",
    gridTemplateColumns: "24px auto",
    justifyContent: "stretch",
    padding: "0",
    // Disable gray background on ListItem hover.
    "&:hover": {
      backgroundColor: "unset"
    }
  },
  facetValueCheckbox: {
    height: "24px",
    width: "24px"
  },
  facetValueNameAndCount: {
    paddingRight: 0
  },
  facetValueName: {
    // Used by end-to-end tests
  },
  facetValueCount: {
    textAlign: "right"
  },
  grayText: {
    color: "silver"
  }
};

class FacetCard extends Component {
  constructor(props) {
    super(props);

    this.facetValues = this.props.facet.values;

    this.totalFacetValueCount = this.sumFacetValueCounts(
      this.props.facet.values,
      []
    );

    this.onClick = this.onClick.bind(this);
    this.isDimmed = this.isDimmed.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.totalFacetValueCount = this.sumFacetValueCounts(
      nextProps.facet.values,
      this.props.selectedValues
    );
  }

  render() {
    const { classes } = this.props;

    const facetValueDivs = this.props.facet.values.map(value => (
      <ListItem
        className={classes.facetValue}
        key={value.name}
        button
        dense
        disableRipple
        onClick={e => this.onClick(value.name)}
      >
        <Checkbox
          className={classes.facetValueCheckbox}
          checked={
            this.props.selectedValues != null &&
            this.props.selectedValues.includes(value.name)
          }
        />
        <ListItemText
          className={classes.facetValueNameAndCount}
          classes={{ primary: this.isDimmed(value) ? classes.grayText : null }}
          primary={
            <div style={{ display: "grid", gridTemplateColumns: "auto 50px" }}>
              <div className={classes.facetValueName}>{value.name}</div>
              <div className={classes.facetValueCount}>{value.count}</div>
            </div>
          }
        />
      </ListItem>
    ));

    return (
      <div className={classes.facetCard}>
        <Typography>{this.props.facet.name}</Typography>
        {this.props.facet.name != "Samples Overview" ? (
          <Typography className={classes.totalFacetValueCount}>
            {this.totalFacetValueCount}
          </Typography>
        ) : null}
        <Typography className={classes.facetDescription}>
          {this.props.facet.description}
        </Typography>
        <List dense className={classes.facetValueList}>
          {facetValueDivs}
        </List>
      </div>
    );
  }

  isDimmed(facetValue) {
    return (
      this.props.selectedValues != null &&
      this.props.selectedValues.length > 0 &&
      !this.props.selectedValues.includes(facetValue.name)
    );
  }

  /**
   * @param facetValues FacetValue[] to sum counts over
   * @param selectedValueNames Optional string[] to select a subset of facetValues to sum counts for
   * @return number count the total sum of all facetValue counts, optionally filtered by selectedValueNames
   * */
  sumFacetValueCounts(facetValues, selectedValueNames) {
    let count = 0;
    if (selectedValueNames == null || selectedValueNames.length === 0) {
      facetValues.forEach(value => {
        count += value.count;
      });
    } else {
      facetValues.forEach(value => {
        if (selectedValueNames.indexOf(value.name) > -1) {
          count += value.count;
        }
      });
    }

    return count;
  }

  onClick(facetValue) {
    // facetValue is a string, eg "female"
    let isSelected;
    if (
      this.props.selectedValues != null &&
      this.props.selectedValues.length > 0 &&
      this.props.selectedValues.includes(facetValue)
    ) {
      // User must have unchecked the checkbox.
      isSelected = false;
    } else {
      // User must have checked the checkbox.
      isSelected = true;
    }

    this.props.updateFacets(
      this.props.facet.es_field_name,
      facetValue,
      isSelected
    );
  }
}

export default withStyles(styles)(FacetCard);
