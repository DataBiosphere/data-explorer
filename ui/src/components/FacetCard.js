import './FacetCard.css';

import React from 'react';
import {Card} from "material-ui/Card";
import {List, ListItem} from "material-ui/List";
import {Checkbox} from "material-ui";

function FacetCard(props) {
    const facet = props.facet;
    const totalCount = props.count;
    const facetValues = facet.values.map((facetValue) =>
        <ListItem
            className="listItem"
            key={facetValue.name}
            leftCheckbox={<Checkbox />}
            primaryText={<div>
                <div className="facetValueName">{facetValue.name}</div>
                <div className="facetValueCount">{facetValue.count}</div>
            </div>}/>
    );
    return (
        <Card className="facetCard">
            <div className="cardHeader">
                <div>{facet.name}</div>
                <div className="subHeader">
                    <span>{totalCount}</span>
                    <span className="numberSelected">0 / {facetValues.length}</span>
                </div>
            </div>
            <List>{facetValues}</List>
        </Card>
    );
}

export default FacetCard;