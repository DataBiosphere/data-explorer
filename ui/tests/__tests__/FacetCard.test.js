import React from "react";
import FacetCard from "../../src/components/facets/FacetCard";
import { getFacet } from "../mockutils";
import { Checkbox, MuiThemeProvider } from "material-ui";
import { getMuiTheme } from "material-ui/styles";
import PropTypes from "prop-types";

test("Renders correctly", () => {
  const tree = shallow(
    <FacetCard facet={getFacet("Facet Name", 3)} updateFacet={() => {}} />
  );

  expect(tree).toMatchSnapshot();
});

test("Updates values after selecting checkboxes", () => {
  const facet = getFacet("Facet Name", 3);
  let mockUpdate = jest.fn();
  let muiTheme = getMuiTheme();
  const tree = mount(<FacetCard facet={facet} updateFacets={mockUpdate} />, {
    context: { muiTheme },
    childContextTypes: { muiTheme: PropTypes.object }
  });
  let checkboxes = tree.find(Checkbox);
  checkboxes
    .first()
    .props()
    .onCheck(null, true);
  checkboxes
    .at(1)
    .props()
    .onCheck(null, true);
  tree.setProps({ facet: facet });

  expect(mockUpdate).toHaveBeenCalledTimes(2);
  expect(tree.find(".totalFacetValueCount").text()).toBe("30");
});

test("Updates values after selecting and unselecting checkboxes", () => {
  const facet = getFacet("Facet Name", 3);
  let mockUpdate = jest.fn();
  let muiTheme = getMuiTheme();
  const tree = mount(<FacetCard facet={facet} updateFacets={mockUpdate} />, {
    context: { muiTheme },
    childContextTypes: { muiTheme: PropTypes.object }
  });
  let checkboxes = tree.find(Checkbox);
  checkboxes
    .first()
    .props()
    .onCheck(null, true);
  checkboxes
    .at(1)
    .props()
    .onCheck(null, true);
  checkboxes
    .first()
    .props()
    .onCheck(null, false);
  checkboxes
    .at(1)
    .props()
    .onCheck(null, false);
  tree.setProps({ facet: facet });

  expect(mockUpdate).toHaveBeenCalledTimes(4);
  expect(tree.find(".totalFacetValueCount").text()).toBe("60");
});
