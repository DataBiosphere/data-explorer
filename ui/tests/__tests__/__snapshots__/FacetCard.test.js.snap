// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<Card
  className="facetCard"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
>
  <div
    className="cardHeader"
  >
    <span>
      Facet Name
    </span>
    <span
      className="totalFacetValueCount"
    >
      60
    </span>
  </div>
  <List>
    <ListItem
      autoGenerateNestedIndicator={true}
      className="listItem"
      containerElement="span"
      disableKeyboardFocus={false}
      disabled={false}
      initiallyOpen={false}
      insetChildren={false}
      key="FacetValue 1"
      leftCheckbox={
        <Checkbox
          disabled={false}
          labelPosition="right"
          onCheck={[Function]}
        />
      }
      nestedItems={Array []}
      nestedLevel={0}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onNestedListToggle={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      open={null}
      primaryText={
        <div
          className=""
        >
          <div
            className="facetValueName"
          >
            FacetValue 1
          </div>
          <div
            className="facetValueCount"
          >
            10
          </div>
        </div>
      }
      primaryTogglesNestedList={false}
      secondaryTextLines={1}
    />
    <ListItem
      autoGenerateNestedIndicator={true}
      className="listItem"
      containerElement="span"
      disableKeyboardFocus={false}
      disabled={false}
      initiallyOpen={false}
      insetChildren={false}
      key="FacetValue 2"
      leftCheckbox={
        <Checkbox
          disabled={false}
          labelPosition="right"
          onCheck={[Function]}
        />
      }
      nestedItems={Array []}
      nestedLevel={0}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onNestedListToggle={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      open={null}
      primaryText={
        <div
          className=""
        >
          <div
            className="facetValueName"
          >
            FacetValue 2
          </div>
          <div
            className="facetValueCount"
          >
            20
          </div>
        </div>
      }
      primaryTogglesNestedList={false}
      secondaryTextLines={1}
    />
    <ListItem
      autoGenerateNestedIndicator={true}
      className="listItem"
      containerElement="span"
      disableKeyboardFocus={false}
      disabled={false}
      initiallyOpen={false}
      insetChildren={false}
      key="FacetValue 3"
      leftCheckbox={
        <Checkbox
          disabled={false}
          labelPosition="right"
          onCheck={[Function]}
        />
      }
      nestedItems={Array []}
      nestedLevel={0}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onNestedListToggle={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      open={null}
      primaryText={
        <div
          className=""
        >
          <div
            className="facetValueName"
          >
            FacetValue 3
          </div>
          <div
            className="facetValueCount"
          >
            30
          </div>
        </div>
      }
      primaryTogglesNestedList={false}
      secondaryTextLines={1}
    />
  </List>
</Card>
`;

exports[`Renders correctly with input 1`] = `
<Card
  className="facetCard"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
>
  <div
    className="cardHeader"
  >
    <span>
      Facet Name
    </span>
    <span
      className="totalFacetValueCount"
    >
      60
    </span>
  </div>
  <List>
    <ListItem
      autoGenerateNestedIndicator={true}
      className="listItem"
      containerElement="span"
      disableKeyboardFocus={false}
      disabled={false}
      initiallyOpen={false}
      insetChildren={false}
      key="FacetValue 1"
      leftCheckbox={
        <Checkbox
          className="listItemCheckbox"
          disabled={false}
          labelPosition="right"
          onCheck={[Function]}
        />
      }
      nestedItems={Array []}
      nestedLevel={0}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onNestedListToggle={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      open={null}
      primaryText={
        <div
          className=""
        >
          <div
            className="facetValueName"
          >
            FacetValue 1
          </div>
          <div
            className="facetValueCount"
          >
            10
          </div>
        </div>
      }
      primaryTogglesNestedList={false}
      secondaryTextLines={1}
    />
    <ListItem
      autoGenerateNestedIndicator={true}
      className="listItem"
      containerElement="span"
      disableKeyboardFocus={false}
      disabled={false}
      initiallyOpen={false}
      insetChildren={false}
      key="FacetValue 2"
      leftCheckbox={
        <Checkbox
          className="listItemCheckbox"
          disabled={false}
          labelPosition="right"
          onCheck={[Function]}
        />
      }
      nestedItems={Array []}
      nestedLevel={0}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onNestedListToggle={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      open={null}
      primaryText={
        <div
          className=""
        >
          <div
            className="facetValueName"
          >
            FacetValue 2
          </div>
          <div
            className="facetValueCount"
          >
            20
          </div>
        </div>
      }
      primaryTogglesNestedList={false}
      secondaryTextLines={1}
    />
    <ListItem
      autoGenerateNestedIndicator={true}
      className="listItem"
      containerElement="span"
      disableKeyboardFocus={false}
      disabled={false}
      initiallyOpen={false}
      insetChildren={false}
      key="FacetValue 3"
      leftCheckbox={
        <Checkbox
          className="listItemCheckbox"
          disabled={false}
          labelPosition="right"
          onCheck={[Function]}
        />
      }
      nestedItems={Array []}
      nestedLevel={0}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onNestedListToggle={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      open={null}
      primaryText={
        <div
          className=""
        >
          <div
            className="facetValueName"
          >
            FacetValue 3
          </div>
          <div
            className="facetValueCount"
          >
            30
          </div>
        </div>
      }
      primaryTogglesNestedList={false}
      secondaryTextLines={1}
    />
  </List>
</Card>
`;
